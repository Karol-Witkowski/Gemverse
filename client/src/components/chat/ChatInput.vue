<template>
  <v-form
    @submit.prevent
    v-model="isFormValid"
  >
    <v-container>
      <v-row>
        <v-col cols="15">
          <v-text-field
            :append-outer-icon="'send'"
            class="input"
            clearable
            clear-icon="cancel"
            @click:append-outer="sendMessage"
            color="primary"
            dense
            :error-messages="inputError"
            label="Message"
            name="message"
            outlined
            v-model="message"
            v-on:keyup="inputError = ''"
            v-on:keyup.enter="sendMessage"
          />
        </v-col>
      </v-row>
    </v-container>
  </v-form>
</template>

<script>
import axios from 'axios';
import { mapGetters } from 'vuex';
import io from 'socket.io-client';

export default {
  name: 'ChatInput',
  data() {
    return {
      inputError: '',
      isFormValid: false,
      message: '',
      socket: io('https://gemverse.herokuapp.com'),
    };
  },

  computed: {
    ...mapGetters(['getCurrentRoom', 'getUserInfo']),
  },

  methods: {
    sendMessage() {
      axios.post(`https://gemverse.herokuapp.com/api/messages/${this.getCurrentRoom.slug}`, {
        message: this.message,
        room: this.getCurrentRoom._id,
        user: this.getUserInfo._id,
      })
        .then((response) => {
          if (response.status === 201) {
            this.socket.emit('sendMessage', response.data.data);
            this.message = '';
          }
        })
        .catch((error) => {
          this.inputError = error.response.data.errors.message.msg;
        });
    },
  },
};
</script>

<style lang="scss">
.v-input__icon {
  font-size: 1.5em!important;
}
</style>
