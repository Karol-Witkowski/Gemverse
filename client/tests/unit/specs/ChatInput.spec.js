import { createLocalVue, mount } from '@vue/test-utils';
import axios from 'axios';
// import io from 'socket.io-client';
import Vue from 'vue';
import Vuetify from 'vuetify';
import ChatInput from '@/components/chat/ChatInput.vue';

const localVue = createLocalVue();
const url = 'http://localhost:3000/api/messages/room-name';
let vuetify;
let wrapper;
const error = {
  response: {
    data: {
      errors: {
        message: {
          msg: 'Message error',
        },
      },
    },
  },
};
const response = {
  data: {
    data: {
      data: 'testData',
    },
    success: true,
  },
};

jest.mock('axios');
vuetify = new Vuetify();

describe('Implementation test for ChatInput.vue - successful HTTP post', () => {
  beforeEach(() => {
    axios.post.mockResolvedValue(response);

    wrapper = mount(ChatInput, {
      localVue,
      mocks: {
        $store: {
          getters: {
            getCurrentRoom: {
              _id:"123testid",
              slug:"room-name",
            },
            getUserInfo: {
              _id:"321testid",
            },
          },
        },
      },
      vuetify,
      data() {
        return {
          inputError: '',
          isFormValid: false,
          message: '',
          // socket: io('http://localhost:3000'),
        };
      },
    });
  });

  afterEach(() => {
    jest.restoreAllMocks();
    wrapper.destroy();
  });

  it('Render correctly', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('Initializes with correct elements', () => {
    // Test buttons initial state
    expect(wrapper.findAll('button').length).toEqual(2);
    expect(wrapper.findAll('button').at(0).text()).toMatch('cancel');
    expect(wrapper.findAll('button').at(1).text()).toMatch('send');

    // Test input initial state
    expect(wrapper.findAll('.v-text-field').length).toEqual(1);
    expect(wrapper.findAll('input').at(0).text()).toEqual('');

    // Test validation initial state
    expect(wrapper.findAll('.v-messages').length).toEqual(1);
    expect(wrapper.findAll('.v-messages').at(0).text()).toEqual('');
    expect(wrapper.vm.isformValid).toBeFalsy();
  });

  it('Enables error messages', async () => {
    await wrapper.setData({
      inputError: 'Input error',
    });

    expect(wrapper.findAll('.v-messages').length).toEqual(1);
    expect(wrapper.findAll('.v-messages').at(0).text()).toEqual('Input error');
  });

  it('Pass validation when message is entered', async () => {
    await wrapper.setData({
      message: 'message',
    });

    // Check that message was entered
    expect(wrapper.findAll('input').at(0).element.value).toEqual('message');
    expect(wrapper.vm.message).toBe('message');

    // Check if validation pass
    expect(wrapper.vm.isFormValid).toBeTruthy();
  });

  it('Should sends post request with message on form submit', async () => {
    await wrapper.setData({
      message: 'message',
    });

    wrapper.vm.sendMessage();

    // Check if post was called
    expect(axios.post).toHaveBeenCalled();

    // Check if post are called once
    expect(axios.post).toHaveReturnedTimes(1);

    // Check if post was called with correct data
    expect(axios.post).toHaveBeenCalledWith(
      url,
      {
        "message": "message",
        "room": "123testid",
        "user": "321testid",
      },
    );
  });
});
