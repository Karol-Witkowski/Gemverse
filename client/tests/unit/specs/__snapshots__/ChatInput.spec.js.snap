// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Implementation test for ChatInput.vue - successful HTTP post Render correctly 1`] = `
<form novalidate="novalidate" class="v-form">
  <div class="container">
    <div class="row">
      <div class="col col-15">
        <div class="v-input input v-input--dense theme--light v-text-field v-text-field--enclosed v-text-field--outlined">
          <div class="v-input__control">
            <div class="v-input__slot">
              <fieldset aria-hidden="true">
                <legend style="width: 0px;"><span>â€‹</span></legend>
              </fieldset>
              <div class="v-text-field__slot"><label for="input-3" class="v-label theme--light" style="left: 0px; position: absolute;">Message</label><input name="message" id="input-3" type="text"></div>
              <div class="v-input__append-inner">
                <div class="v-input__icon v-input__icon--clear"><button disabled="disabled" type="button" aria-label="clear icon" class="v-icon notranslate v-icon--disabled v-icon--link material-icons theme--light">cancel</button></div>
              </div>
            </div>
            <div class="v-text-field__details">
              <div class="v-messages theme--light">
                <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
              </div>
            </div>
          </div>
          <div class="v-input__append-outer">
            <div class="v-input__icon v-input__icon--append-outer"><button type="button" aria-label="append icon" class="v-icon notranslate v-icon--link material-icons theme--light">send</button></div>
          </div>
        </div>
      </div>
    </div>
  </div>
</form>
`;
